library/std/src/io/error.rs

```diff
@@ -127,6 +133,35 @@ pub enum ErrorKind {
     /// requested to not occur.
     #[stable(feature = "rust1", since = "1.0.0")]
     WouldBlock,
+    /// A filesystem object is, unexpectedly, not a directory.
+    ///
+    /// For example, a filesystem path was specified where one of the intermediate directory
+    /// components was, in fact, a plain file.
+    #[stable(feature = "io_error_unix_more", since = "1.51.0")]
+    NotADirectory,
+    /// The filesystem object is, unexpectedly, a directory.
+    ///
+    /// A directory was specified when a non-directory was expected.
+    #[stable(feature = "io_error_unix_more", since = "1.51.0")]
+    IsADirectory,
+    /// A non-empty directory was specified where an empty directory was expected.
+    #[stable(feature = "io_error_unix_more", since = "1.51.0")]
+    DirectoryNotEmpty,
+    /// The filesystem is read-only, but a write operation was attempted.
+    #[stable(feature = "io_error_unix_more", since = "1.51.0")]
+    ReadOnlyFilesystem,
+    /// Too many levels of symbolic links.
+    ///
+    /// There was a symbolic link loop; or, the system-specific limit on the depth of symlink
+    /// traversal was exceded.
+    #[stable(feature = "io_error_unix_more", since = "1.51.0")]
+    SymbolicLinkLoop,
```

Comment:
I don't think Windows has an equivalent of this. There's [`ERROR_STOPPED_ON_SYMLINK`](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--500-999-), [`ERROR_SYMLINK_CLASS_DISABLED`](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--1300-1699-), and [`ERROR_SYMLINK_NOT_SUPPORTED`](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--1300-1699-), but they don't seem to match this exact case

Reply:
OK

Reply:
While excessive or cyclic symbolic links are the most common scenario in which this will arise, there are other drivers and subsystems which can return `ELOOP`. For instance, if you attempt to execute a binary on Linux that's using `binfmt_misc`, and that ends up recursing (because the interpreter needs interpreting) too many times, you get `ELOOP`. Or if you use epoll and try to put the epoll fd inside itself. Or if your network routing has a cycle. In general, Linux tends to use `ELOOP` for any case of "I'm refusing to follow a cycle this deeply inside the kernel", not just for symlinks.

I think, at a minimum, it'd be helpful to include a catch-all in the documentation for this: "or a similar kind of unexpected cycle or excessively deep recursion occurred".

I'm also wondering if it'd be appropriate to leave "SymbolicLink" out of this name, and just call it "Loop". I don't want to make this less descriptive, but I think this is one case where it'd be actively *confusing* to assume this is always about symlinks.

Reply:
> While excessive or cyclic symbolic links are the most common scenario in which this will arise, there are other drivers and subsystems which can return ELOOP

Hmmm.  Good point.

I have renamed it `FilesystemLoop` and made the documentation more vague :-).

For now you can see that change in 1d9c4ffcd5a whjich I have just pushed, but I am going to rebase this again when I'm done folding in changes etc.

Reply:
I think an even more generic name would be appropriate. Maybe something like `TooMuchRecursion` or `IndirectionLimit`

E.g. the epoll_ctl(2) use has nothing to do with filesystems at all

```
       ELOOP  fd refers to an epoll instance and this EPOLL_CTL_ADD
              operation would result in a circular loop of epoll
              instances monitoring one another or a nesting depth of
              epoll instances greater than 5.
```

Reply:
I agree, and `IndirectionLimit` seems like a good generic name.

Reply:
I have updated the description to include this (and will push that shortly).

I don't think the name should be changed.  Because of the way Unix reuses errno values, it is often going to be the case that there will be use cases which don't fit the primary summary.  These are typically less common (and less user facing) than the older or more primary meaing.

I think trying to capture all these uses of every errno value would result in distressingly vague names for `ErrorKind`.  Sometimes, even, the uses are quite incoherent.  (Consider `EAGAIN` from `fork` vs `read`.)

I think it is better to capture this information in the extended description.  Also we agreed to add to the documentation information about which errno values map to which `ErrorKind`s.  That will help too.

In the case of `ELOOP`, I think it is important to retain a good experience for an end user who uses some Rust program which finds a symbolic link loop, and print a message and an abbrevation that is meaningful to them.  And it is important that Rust programmers who are writing generic file-handling code can easily find the symbolic link loop error code.

Conversely `ELOOP` from `epoll_ctl` will generally be experienced by an event loop, and be fatal.  It's also, I guess, really quite rare.